src/mpi/coll/op/opequal.c

```diff
@@ -0,0 +1,241 @@
+                a[i+1] = 1;                             \
+            }                                           \
+        }                                               \
+    }                                                   \
+    break
+
+void MPIR_EQUAL_user_defined_datatype_compare(void *invec, void *inoutvec, int *Len, MPI_Datatype * type)
+{
+	int mpi_errno = MPI_SUCCESS;
+	int i, j, size, len = *Len, is_equal; 
+	int data_len = 0, type_len = 0;
+	int num_ints, num_adds, num_types, combiner, *ints;
+	void *invec_i_pos, *invec_i_bool_pos, *inoutvec_i_pos, *inoutvec_i_bool_pos;
+	MPI_Aint *adds = NULL;
+	MPI_Aint lb, extent;
+	MPI_Datatype *types;
+
+	 /* decode */
+	MPIR_Type_get_envelope(*type, &num_ints, &num_adds, &num_types, &combiner);
+
+	if(num_types < 2 || combiner != MPI_COMBINER_STRUCT) /*At least 2 elements is required, data, result*/
+	{
+		MPIR_ERR_SET1(mpi_errno, MPI_ERR_OP, "**opundefined", "**opundefined %s", "MPIX_EQUAL");
+		return;
+	}
+
+	ints = (int *)malloc(num_ints * sizeof(*ints));
+
+	if (num_adds || (num_ints != num_adds+1))/*ints[0] is the length*/
+	{
+	    adds = (MPI_Aint *)malloc(num_adds * sizeof(*adds));
+	}
+	else /*adds is required to avoid impacts of struct alignment*/
+	{
+		MPIR_ERR_SET1(mpi_errno, MPI_ERR_OP, "**opundefined", "**opundefined %s", "MPIX_EQUAL");
+		return;
+	}
+
+	types = (MPI_Datatype *)malloc(num_types * sizeof(*types));
+
+	mpi_errno = MPIR_Type_get_contents(*type, num_ints, num_adds, num_types, ints, adds, types);
+
+	if(types[num_types-1] != MPI_INT) /*The last element has to be int*/
+	{
+		MPIR_ERR_SET1(mpi_errno, MPI_ERR_OP, "**opundefined", "**opundefined %s", "MPIX_EQUAL");
+		return;
+	}
+
+	MPIR_Type_get_extent_impl(*type, &lb, &extent);
+	type_len = extent - lb;
+
+	for (i = 0; i < len; ++i)    	
+	{
+		invec_i_pos = invec + i*type_len;
+		invec_i_bool_pos = invec_i_pos + adds[num_adds-1];
+		inoutvec_i_pos = inoutvec + i*type_len;
+		inoutvec_i_bool_pos = inoutvec_i_pos + adds[num_adds-1];
+		if((*(int*)invec_i_bool_pos == 0) ||
+			(*(int*)inoutvec_i_bool_pos == 0))
+		{
+			*(int*)inoutvec_i_bool_pos  = 0;
+		}
+		else
+		{/*compare the content of the struct*/
+			is_equal = 1;
+			for(j = 0; j < num_adds-1; ++j)
+			{
+				MPI_Type_size(types[j], &size); 
+				data_len = ints[j + 1] * size;
+
+				if(memcmp(invec_i_pos+adds[j], inoutvec_i_pos+adds[j], data_len))
+					is_equal = 0;
+			}
+
+			*(int*)inoutvec_i_bool_pos  = is_equal;
+		}
+	}
+}
+
+void MPIR_EQUAL(void *invec, void *inoutvec, int *Len, MPI_Datatype * type)
+{
+    int i, len = *Len;
+
+#ifdef HAVE_FORTRAN_BINDING
+#ifndef HAVE_NO_FORTRAN_MPI_TYPES_IN_C
+    int flen = len * 2;         /* used for Fortran types */
+#endif
+#endif
+    
+    switch (*type) {
+            /* first the C types */
+        case MPI_2INT:
+            MPIR_EQUAL_C_CASE_INT(MPIR_2int_eqltype);
+        case MPI_FLOAT_INT:
+            MPIR_EQUAL_C_CASE_FLOAT(MPIR_floatint_eqltype);
```

Comment:
Remove the float support, bring explicit `break` here.

Reply:
Removed.